assume/common/units_operator.py

```diff
@@ -290,28 +301,67 @@ async def formulate_bids(
         """
         sends the current rl_strategy update to the output agent
         """
-
-        unit_rl_strategy_dfs = []
-        for unit_id, unit in self.units.items():
-            # rl only for energy market for now!
-            if isinstance(unit.bidding_strategies.get("energy"), LearningStrategy):
-                data = pd.DataFrame(
-                    {
-                        "profit": unit.outputs["profit"].loc[start],
-                        "reward": unit.outputs["reward"].loc[start],
-                        "regret": unit.outputs["regret"].loc[start],
-                    },
-                    index=[start],
+        learning_unit_count = 0
+        action_dimension = 0
+        for unit in self.units.values():
+            bidding_strategy = unit.bidding_strategies.get(marketconfig.product_type)
+            if isinstance(bidding_strategy, LearningStrategy):
+                learning_unit_count += 1
+                # should be the same across all strategies
+                action_dimension = bidding_strategy.act_dim
+        if learning_unit_count > 0:
+            start = orderbook[0]["start_time"]
+            unit_rl_strategy_dfs = []
+
+            all_observations = []
+            all_rewards = []
+            try:
+                import torch as th
+
+                all_actions = th.zeros(
+                    (learning_unit_count, action_dimension), device="cpu"
                 )
-                data["unit"] = unit_id
-                unit_rl_strategy_dfs.append(data)
-
-        if len(unit_rl_strategy_dfs):
+            except ImportError:
+                logger.error(
+                    "tried writing learning_params, but torch is not installed"
+                )
+                all_actions = np.zeros((learning_unit_count, action_dimension))
+
+            i = 0
+
+            for unit_id, unit in self.units.items():
+                # rl only for energy market for now!
+                if isinstance(
+                    unit.bidding_strategies.get(marketconfig.product_type),
+                    LearningStrategy,
+                ):
+                    data = pd.DataFrame(
+                        {
+                            "profit": unit.outputs["profit"].loc[start],
+                            "reward": unit.outputs["reward"].loc[start],
+                            "regret": unit.outputs["regret"].loc[start],
+                        },
+                        index=[start],
+                    )
+                    data["unit"] = unit_id
+                    unit_rl_strategy_dfs.append(data)
+                    all_observations.append(unit.outputs["rl_observations"][start])
+                    all_actions[i, :] = unit.outputs["rl_actions"][start]
+                    all_rewards.append(unit.outputs["rl_rewards"][start])
+
+                    i += 1
+            all_observations = np.array(all_observations)
```

Comment:
Here, I have an error because all_observations consists of several tensors.
If line 452 is changed to "all_observations.append(np.array(unit.outputs["rl_observations"][start]))" it runs without error.

Reply:
Thats somewhat weird because we are converting it to an array in 357 but if this is needed, thats finde